Hide embargoed versions assets

dchiquito opened the pull request:
I missed a spot when concealing the `DandisetViewSet` from returning data on embargoed Dandisets: Versions and Assets would still dispense data happily even when associated with embargoed Dandisets. This fixes that, and adds tests.

Technically, this implementation leaks the fact that an embargoed dandiset exists with that identifier, since querying `/api/dandisets/{embargoed_dandiset_id}/versions/` will return an empty pagination instead of a 404. I think that's fine, since embargoed dandisets will conspicuously occupy gaps in the dandiset identifier sequence, and there isn't anything a bad actor can do with an identifier. @waxlamp can you advise?

Fixes #667

yarikoptic commented:
FWIW, leaking knowledge about embargoed identifier existence is IMHO ok. And for that reason I even wonder if we should step away from GitHub inspired 404 into a proper 401 and 403 responses depending on either session is already authenticated or not.

But also I wondered if there is a way in Django to provide handling for that `/api/dandisets/{embargoed_dandiset_id}/` prefix, so that it is always invoked first for any end point starting with it? If so, then all embargo authentication related handling at dandiset level could be conveniently centralized

yarikoptic commented:
BTW, forgot to ping @satra on above

satra commented:
identifier existence is ok.

waxlamp commented:
> FWIW, leaking knowledge about embargoed identifier existence is IMHO ok. And for that reason I even wonder if we should step away from GitHub inspired 404 into a proper 401 and 403 responses depending on either session is already authenticated or not.
> 

I agree with this. See #669. I'm deferring to @yarikoptic and @satra on the issue of "leaking' the existence of embargoed dandisets--my own $0.02 is that because this is an open scientific archive, there's nothing to be lost and perhaps something small to be gained by subtly broadcasting the existence of these dandisets.

I would possibly go even further and imagine an eventual feature that would enable information about embargoed data to be published, without publishing the data itself. But this is just me thinking out loud.

> But also I wondered if there is a way in Django to provide handling for that `/api/dandisets/{embargoed_dandiset_id}/` prefix, so that it is always invoked first for any end point starting with it? If so, then all embargo authentication related handling at dandiset level could be conveniently centralized

I don't think this is how it would be done, but rather with the right permission-related decorators on all relevant endpoints. I'll leave the details of this to @dchiquito.

yarikoptic commented:
> > But also I wondered if there is a way in Django to provide handling for that `/api/dandisets/{embargoed_dandiset_id}/` prefix, so that it is always invoked first for any end point starting with it? If so, then all embargo authentication related handling at dandiset level could be conveniently centralized
> 
> I don't think this is how it would be done, but rather with the right permission-related decorators on all relevant endpoints. I'll leave the details of this to @dchiquito.

I do get such an approach.  The question was more of satisfying my curiosity on either a more robust (e.g. so a forgotten decorator on a new endpoint doesn't leak it to public) way possible.

waxlamp commented:
> > > But also I wondered if there is a way in Django to provide handling for that `/api/dandisets/{embargoed_dandiset_id}/` prefix, so that it is always invoked first for any end point starting with it? If so, then all embargo authentication related handling at dandiset level could be conveniently centralized
> > 
> > 
> > I don't think this is how it would be done, but rather with the right permission-related decorators on all relevant endpoints. I'll leave the details of this to @dchiquito.
> 
> I do get such an approach. The question was more of satisfying my curiosity on either a more robust (e.g. so a forgotten decorator on a new endpoint doesn't leak it to public) way possible.

There is probably some way to do what you're suggesting, but I don't think it would be the "django way", but rather a more generic Python approach to refactoring some gating logic out or something. It is a good question though, about triggering common logic because of the common endpoint prefix.

dchiquito commented:
I implemented it in the most generic method I know how.

To give an overview, we are using a [library](https://www.django-rest-framework.org/) that lets us define a class that represents a family of endpoints. On that class, we define a `queryset` or a `get_queryset` method that returns a Django QuerySet, which is directly translatable into SQL by the Django ORM. I modified the `queryset` on `VersionViewSet` and `AssetViewSet` to filter out dandisets which are not visible to the current user.
The issue with implementing it in a more generic way is that the construction of the `queryset` directly impacts the SQL performance. In general I think it's better to be explicit when defining a `queryset` so that it's easier to read and debug. We are also unlikely to be adding any new models like `Version` or `Asset` that might belong to embargoed dandisets that are similar enough to be able to use some kind of generic parent class.
